### ISSUE TITLE
[csrng] Define how CSRNG status response should operate

### ISSUE THREAD

**Issue by marnovandermaas** [lowRISC/opentitan#16516]
With the current RTL, the `sw_cmd_sts.cmd_sts` is never high and the bits `hw_exc_sts` are also never high. This is due to the current design of the RTL not being designed to pass back anything but a successful status. For example, when an illegal command is received the main state machine simply hangs until it is reset and does not return any status.

In the documentation it states that there are three reasons for there to be an illegal status: 
1. https://github.com/lowRISC/opentitan/blob/b15fa805be3bae4528310e614372d91761faef28/hw/ip/csrng/doc/_index.md?plain=1#L416
1. https://github.com/lowRISC/opentitan/blob/b15fa805be3bae4528310e614372d91761faef28/hw/ip/csrng/doc/_index.md?plain=1#L417
1. https://github.com/lowRISC/opentitan/blob/b15fa805be3bae4528310e614372d91761faef28/hw/ip/csrng/doc/_index.md?plain=1#L418

We should decide whether the current design is desirable and verify that each of these scenarios have the correct behavior.

Assuming no changes to RTL:
> estimate 2

**Comment by marnovandermaas** [comment]
This issue is also related to: https://github.com/lowRISC/opentitan/issues/16499

**Comment by vogelpi** [comment]
Thanks for filing this @marnovandermaas and @ctopal . I agree with you, we should discuss how the RTL should behave and then also verify that. I don't think any of this is super critical for ES but it seems that there is at least the need for some clarification. Maybe it's all about aligning the documentation, maybe some RTL changes should be done. It would be best to discuss this with @mwbranstad after the break to get his view and the original intention. 

Regarding the second reason for an failure status "Attempts to use an instance which has not been properly instantiated" I think I've seen something in the RTL but it was more checking the RTL instantiation of the command stage, i.e., SV parameters. From the spec, I would have expected the design to report a failing status if e.g. a generate command is received without prior instantiate command.

We've also discussed to return a failing status bit upon a fatal alert but that wouldn't be in line with how we handle fatal alerts in other places in the system and it might need big design changes. But it could be useful to get some more immediate feedback in the case of misconfiguration. E.g. if an EDN or SW provides an illegal command, there is no response and CSRNG and downstream modules just hang. Software will have to figure out what caused this. This could maybe be simplified if also EDN/SW would see the failure via `sts` bit more directly. Let's discuss with Mark after the break.

**Comment by vogelpi** [comment]
I just noted that the same question was brought up by @cindychip for EDN in #15894:
> On CSRNG side, (only from reading the spec, I do not actually know the simulation results). It seems like CSRNG will fire an interrupt.

@mwbranstad explained
> the only case where the sts bit will be on is when a silicon error occurs, or some attacker is driving this bit on. In other words there is no functional reason that this bit will set. I would probably rate this bit being set as more severe, such as a fatal alert.

and @tjaychen agreed
>  if it's alright with you, i can go in and make that change.
To not disturb things further, we can leave the interrupt and also add the fatal indication.
What do you think?

In general, my feeling is that fatal errors should not be factored into the sts error signal. Because these are a very strong indication of an attack going on. They're thus fatal and the chip needs to be rebooted anyway. I see little use in reporting this condition further downstream.

But I would consider it useful if CSRNG would share information about configuration errors and misuse e.g. through software. Such things are usually signaled as recoverable alerts. For example, if EDN forwards an invalid command to CSRNG it currently gets no response at all, i.e., it just hangs. CSRNG doesn't acknowledge the command but it signals a recoverable alert and its main FSM halts. The other scenarios mentioned by @marnovandermaas (exceeding seed life, requests to non-instantiated CSRNG instances, entropy_src errors) go in the same direction. In most cases they are due to software not using CSRNG in the correct/intended way.

**Comment by andreaskurth** [comment]
Triaged for `csrng`. Assigning to M2.5 with https://github.com/lowRISC/opentitan/labels/Priority%3AP1 because we need to take a decision on this, as any change will require time for implementation and verification.

**Comment by marnovandermaas** [comment]
Assuming we decide not to make changes to the RTL:
> estimate 2

**Comment by vogelpi** [comment]
We've discussed this in the Security WG meeting on Mar 30 2023 and concluded the following:

- In case CSRNG detects an internal error condition (e.g. invalid command), its main FSM stops and signals a recoverable alert. It stops serving entropy to any interfaces eventually. As a result of that, entropy consumers will stall at some point and software will note this. Software then has to disable and re-enable the entropy complex to get things going again.
- All known cases that could lead to such an error condition (if the design was changed as discussed in this issue) are due to mis-use/misconfiguration by software. While it would be nice to get more direct feedback for this mis-use/config to ease debugging, the stalling is eventually visible to software. And there exists a way to move out of such a situation.
- Given the expected effort and our deadlines, we have agreed to leave this part of the design as is for M2.5.

However, it is also understood that this should be fixed for a future release. I am thus changing the labels accordingly.
